router-tests/websocket_test.go

```diff
@@ -2159,3 +2166,236 @@ func expectConnectAndReadCurrentTime(t *testing.T, xEnv *testenv.Environment) {
+
+		t.Parallel()
+
+		totalUpdates := 5
+
+		// Channel to collect pings from the router
+		pingReceived := make(chan bool, totalUpdates)
+
+		// Channel for handling websocket writes
+		wsWriteCh := make(chan wsMessage)
+
+		// Middleware to handle WebSocket connections
+		wsMiddleware := func(handler http.Handler) http.Handler {
+			// Configure the WebSocket upgrader
+			upgrader := websocket.Upgrader{
+				CheckOrigin:  func(r *http.Request) bool { return true },
+				Subprotocols: []string{"graphql-transport-ws"},
+			}
+
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				// Upgrade the connection
+				conn, err := upgrader.Upgrade(w, r, nil)
+				if err != nil {
+					t.Logf("WebSocket upgrade failed: %v", err)
+					return
+				}
+				defer conn.Close()
+
+				// Start a goroutine to handle all writes to the websocket
+				go func() {
+					for msg := range wsWriteCh {
+						err := conn.WriteJSON(msg.data)
+						msg.done <- err
+					}
+				}()
+
+				// Handle the GraphQL protocol
+				for {
+					// Read message
+					messageType, message, err := conn.ReadMessage()
+					if err != nil {
+						// Normal close is OK
+						if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+							return
+						}
+						t.Logf("Failed to read message: %v", err)
+						return
+					}
+
+					// Only handle text messages
+					if messageType != websocket.TextMessage {
+						continue
+					}
+
+					// Parse the message
+					var msg map[string]interface{}
+					if err := json.Unmarshal(message, &msg); err != nil {
+						t.Logf("Failed to unmarshal message: %v", err)
+						continue
+					}
+
+					// Process based on message type
+					msgType, ok := msg["type"].(string)
+					if !ok {
+						continue
+					}
+
+					switch msgType {
+					case "connection_init":
+						// Acknowledge connection
+						done := make(chan error, 1)
+						wsWriteCh <- wsMessage{
+							data: map[string]string{"type": "connection_ack"},
+							done: done,
+						}
+						if err := <-done; err != nil {
+							t.Logf("Failed to send connection_ack: %v", err)
+							return
+						}
+
+					case "ping":
+						// Record ping and send pong
+						select {
+						case pingReceived <- true:
+							t.Log("Received ping from router")
+						default:
+							// Don't block if channel is full
+						}
+
+						done := make(chan error, 1)
+						wsWriteCh <- wsMessage{
+							data: map[string]string{"type": "pong"},
+							done: done,
+						}
+						if err := <-done; err != nil {
+							t.Logf("Failed to send pong: %v", err)
+							return
+						}
+
+					case "subscribe":
+						// Handle countEmp subscription
+						if payload, ok := msg["payload"].(map[string]interface{}); ok {
+							if query, ok := payload["query"].(string); ok && strings.Contains(query, "countEmp") {
+								go handleCountEmpSubscription(t, wsWriteCh, msg["id"], 200*time.Millisecond, totalUpdates)
+							}
+						}
```

Comment:
The `msg` type seems like it would be fairly easy to mostly if not fully unmarshal into a concrete struct, is that not possible? I believe it would make the test easier to read.

Reply:
It is possible. Although it doesn't add much value and I see your point and fixed it.